design-proposals/amd-sev.md

```diff
@@ -0,0 +1,253 @@
+
+```xml
+<domain>
+  ...
+  <launchSecurity type='sev'>
+    <policy>0x0001</policy>
+    <cbitpos>47</cbitpos>
+    <reducedPhysBits>1</reducedPhysBits>
+    <dhCert>RBBBSDDD=FDDCCCDDDG</dhCert>
+    <session>AAACCCDD=FFFCCCDSDS</session>
+  </launchSecurity>
+  ...
+</domain>
+```
+
+Here the `policy` parameter allows setting the guest policy flags. The optional `dhCert` and `session` provide the guest owner's base64 encoded `DH` (Diffie-Hellman) key and the guest owner's base64 encoded `launch blob` respectively (those are needed for the attestation).
+
+Additionally to start a SEV VM there is a need to provide `cbitpos` and `reducedPhysBits`. The parameters are hypervisor-dependent and can be obtained from the `<sev></sev>` element of the `virsh domcapabilities`:
+
+```xml
+<domainCapabilities>
+...
+  <features>
+    ...
+    <sev supported='yes'>
+      <cbitpos>47</cbitpos>
+      <reducedPhysBits>1</reducedPhysBits>
+    </sev>
+    ...
+  </features>
+</domainCapabilities>
+```
+
+**Note:** in recent libvirt `cbitpos` and `reducedPhysBits` can be ommited.
+
+Apart from setting the parameters, there are several more prerequisites to run SEV guests:
+
+- all virtio devices need to be configured with the `iommu='on'` attribute in their `<driver>` configuration
+- all memory regions used by the VM must be `locked` for Direct Memory Access (DMA) and to prevent swapping
+
+From the implementation perspective the required parameters can be mapped to the VMI spec:
+
+```yaml
+spec:
+  domain:
+    launchSecurity:
+      sev: {}
+```
+
+Additional logic and validation shall be added to ensure that the domain xml meets the prerequisites. Apart from that `/dev/sev` device needs to be exposed to the VM pod with proper access bits (currently QEMU requires `read-write` access).
+
+SEV-capable nodes should be properly labeled. Then a corresponding node selector can be added to VM pods so they are scheduled correctly.
+
+## Attestation
+
+[Attestation](https://github.com/AMDESE/sev-tool#proposed-provisioning-steps) is an interactive process between `platform owner` and `guest owner`. It is carried out to establish a certain level of trust. To leverage the benefits of memory encryption a guest owner provides an encrypted VM disk. To decrypt the data and to run a VM with such a disk there is a need to `inject a secret` into the VM when the system is about to boot. The secret is known only by the guest owner. The goal of the attestation process is to validate that the platform is genuine and it really runs a non-compromised SEV hardware. Also it is needed to establish a private communication channel with the platform so the guest owner can securely provide the secret. Attestation consists of the following steps:
+
+- VM gets scheduled on a SEV-capable node
+- guest owner validates the platform on that specific node
+- guest owner establishes a secure channel with the platform by providing the session parameters
+- VM is started in a paused state
```

Comment:
The multi-stage dance where the mgmt app has to separate VM startup into two phases, with user interaction in between scheduling a VM to a specific host and then later starting it on this host, is one of the major pain points in implementing support for measured boot with SEV.  It makes the SEV VM launch flow significantly different from a normal VM startup where scheduling the VM on a host and launching the VM are essentially part of the same logic operation.

The further complexity where by mgmt apps have to expose extra information about the host SEV capabilities, the guest measurement and add APIs fo secret injection are a further burden on the support for SEV VMs.

It further creates a burden on the guest owner, because to run trusted VMs, they need to use an attestation agent and this agent now needs to implement support for the different APIs exposed by many different cloud vendor, to avoid vendor lockin.

To try to address this extra burden on mgmt apps and users I have raised the question of whether libvirt can facilitate automation for almost all of this attestation related extra work. It would reduce the problem of enabling SEV attestation to merely providing a URL pointing to the attestation agent controlled by the guest owner, with libvirt then directly talking to this agent. 

https://listman.redhat.com/archives/libvir-list/2022-February/msg00857.html

With such a proposal the implementation of AMD SEV in KubeVirt would be significantly simpler IMHO, so would appreciate feedback from KubeVirt maintainers on whether this idea is appealing.

Reply:
I think the proposed approach with handling attestation in libvirt is definitely valid. It may indeed simplify the implementation  here. KubeVirt will not need to expose all the APIs for handling the attestation. Though it will not eliminate the "two phases" (i.e. 1) scheduling on a node, verifying the certs, preparing the launch blob and then 2) launching paused, measuring, injecting secret, upause) but all that will happen behind the scenes without the need to involve the guest owner. 

The only thing that bothers me in that flow is the attestation server and how it fits into the SEV trust model. The guest owner will need to share the secret with this service. So I would assume it should be controlled solely by the guest owner. Maybe even running locally?... Just thoughts...

Anyway, I the idea with libvirt handling the attestation seems good to me. Would be great to see that moving forward.

Maybe other thought/opinions here?

Reply:
> Though it will not eliminate the "two phases" (i.e. 1) scheduling on a node, verifying the certs, preparing the launch blob and then 2) launching paused, measuring, injecting secret, upause) but all that will happen behind the scenes without the need to involve the guest owner.

It would absolutely remove eliminate it from KubeVirt's POV too. It merely becomes a normal 'virDomainCreate' API call for KubeVirt as with any other non-SEV VM.   Libvirt simply blocks the virDomainCreate API impl from spawning QEMU until it has received the launch blob & certs back from the attestation agent.

> The only thing that bothers me in that flow is the attestation server and how it fits into the SEV trust model. The guest owner will need to share the secret with this service. So I would assume it should be controlled solely by the guest owner.

Yes the attestation server is something that has to be trusted by the guest owner. This means running it somewhere the guest owner considers to be trustworthy, which either means it has to be outside the untrustworthy cloud, or they need to somehow bootstrap the trust relationship for a 1st VM and then run their agent in this VM. The latter approach would mean that the cloud would actually still need to expose the info about SEV in its APIs, which is kinda defeats some of the benefits.

Reply:
Raised this in the KubeVirt mailing list https://groups.google.com/g/kubevirt-dev/c/a6XlT0qRNAY to have some attention from a broader audience.

Reply:
I would love to see some improvements on the flow.

> > The only thing that bothers me in that flow is the attestation server and how it fits into the SEV trust model. The guest owner will need to share the secret with this service. So I would assume it should be controlled solely by the guest owner.
> 
> Yes the attestation server is something that has to be trusted by the guest owner. This means running it somewhere the guest owner considers to be trustworthy, which either means it has to be outside the untrustworthy cloud, or they need to somehow bootstrap the trust relationship for a 1st VM and then run their agent in this VM. The latter approach would mean that the cloud would actually still need to expose the info about SEV in its APIs, which is kinda defeats some of the benefits.

Here are some potential pitfalls if we go down that path:

 * As already mentioned, a public accessible trusted place needs to exist:
   What if we deal with an air-gaped installation? There one typicall has access to clusters via e.g. a bastion, or with other means, typically going one direction (kubectl can reach the cluster, but not necessarily the other way round).
 * Libvirt has no network access in the kubevirt design. We could somehow work around that by routing through virt-handler, but having libvirt trying to reach external places, is not ideal from a kubevirt architecture

In general, from a k8s architecture perspective, if we talk about automating flows like this, a controller model may fit better. The controller watches the state of the VMI, goes through the atestation process and approves the start without further user interaction. The benefit of this is, that controllers only need access to the cluster in a similar way like `kubectl`. They can be deployed in the cluster (limited trust, but may still help to help preventing breaches on individual nodes), or it can be deployed anywhere on the world.

I understand that this would eliminate the benefit of a common protocol. However, if libvirt would decide to implement such a REST-based approach, we could still make use of it internally, or still use the other libvirt client calls, so we would not be in the way. I am just not sure if the reach-out approach from libvirt is the right pattern to be useful for all kinds of cloud-services and scenarios.

Reply:
> Libvirt has no network access in the kubevirt design.

Yes but there are already a couple of use-cases where KubeVirt proxies the data transfer: console/VNC or live migration. Communication with the atestation service can be proxied as well. In that case it should be sufficient to have access to the cluster via kubectl/virtctl. Or do you think it may have some implications?

> The controller watches the state of the VMI, goes through the atestation process and approves the start without further user interaction.

I think in case of the such a model a guest owner will have to share the secret with the controller. That in my opinion does not fit well with the idea behind SEV where a guest owner does not want to expose the secrets to the cloud provider/infra. I hardly see how it can be done without user interaction. Only maybe if there is an "attestation service" that will talk on behalf of the user with that controller. Or did I understand the idea wrong?

Reply:
> > Libvirt has no network access in the kubevirt design.
> 
> Yes but there are already a couple of use-cases where KubeVirt proxies the data transfer: console/VNC or live migration. Communication with the atestation service can be proxied as well. In that case it should be sufficient to have access to the cluster via kubectl/virtctl. Or do you think it may have some implications?

Yes that would be possible. However having the need to create port-forwards over the place til it reaches the user looks like not the right solution.


> > The controller watches the state of the VMI, goes through the atestation process and approves the start without further user interaction.
> 
> I think in case of the such a model a guest owner will have to share the secret with the controller. That in my opinion does not fit well with the idea behind SEV where a guest owner does not want to expose the secrets to the cloud provider/infra. I hardly see how it can be done without user interaction. Only maybe if there is an "attestation service" that will talk on behalf of the user with that controller. Or did I understand the idea wrong?

I agree with what you say. All I meant is, that a controller-like pattern on the user-side (or trusted remote side) may suite better than letting libvirt reach out. I just used `controller` in the sense of describing the pattern. This could for instance just be part of virtctl. At the end with some tool (even if it is just on the machine of the user) the secret has to be exposed.

Reply:
For more context, while I am not necessarily promoting the idea of creating additional k8s objects for SEV, it has some similarities to certificate signing request flows in k8s, where certain things, like joining new nodes, are blocked until someone or something examines the CSR and approves it explicitly.

Reply:
> - As already mentioned, a public accessible trusted place needs to exist:
>    What if we deal with an air-gaped installation? There one typicall has access to clusters via e.g. a bastion, or with other means, typically going one direction (kubectl can reach the cluster, but not necessarily the other way round).
> - Libvirt has no network access in the kubevirt design. We could somehow work around that by routing through virt-handler, but having libvirt trying to reach external places, is not ideal from a kubevirt architecture

So when I say that libvirt would be able to connect to an attestation server, it doesn't necessarily require that libvirt have a direct outbound connection to the server. KubeVirt is not so unusual in not allowing the compute host to have arbitrary outbound network access, so it is a general concern. In KubeVirt case libvirt is admittedly even more restricted than some other clouds, but it isn't a show stopper.

Another concern is that the user may not have somewhere to run their attestation server that allows inbound network access. eg if running it on your laptop it can be behind NAT, so k8s would not be able to reach it.

A way to address all these issues would be for the cloud infrastructure to provide a transparent proxy that sits in between libvirt and the end user's attestation server. The proxy could sit on the internet perimeter of the cloud infra, such that the end users attestation service can establish a connection to it, and then libvirt can also establish a connection to it from the compute host, then it just forwards traffic in both directions. This would avoid the NAT issue for a guest owner running a remote attestation service, and also solve the issue of libvirt not having general outbound access. The KubeVirt case might involve multiple hops for the proxying. Ultimately this isn't all that different from exposing QEMU's VNC console, aside from the fact that libvirt reaches out, rather than a user reaching in eg we're just reversing one arrow in the comparison of:

   user -> proxy -> vnc

vs

  user  -> proxy <- libvirt SEV

I very much doubt the attestation service protocol would be anything tied to libvirt explicitly, as its design is really driven by what information needs to be provided by the two end points. So if we get a standardized protocol for attestation, it is certainly possible for KubeVirt to act as the endpoint on the cloud side, and make libvirt API calls to complete the work, if desired. 

Finally one thing to bear in mind is that what we're considering here has primarily only taken into account AMD SEV and AMD SEV-ES since these work in essentially the same way, the only difference being extra CPU register state in the SEV-ES case. 

With the newer AMD SEV-SNP the architecture is significantly changed. Instead of attestation being initiated on the host before guest CPUs run, with SNP it is initiated inside the guest after guest CPUs run. With such  a scheme attestation needs to take place as soon as possible during guest boot. With even the initrd arguably being inconveniently late, it is quite possible it will be triggered from inside the guest firmware (OVMF) before it loads the kernel. The guest firmware is can't be assumed to have network access to reach the guest owner's attestation service, so it is highly likely it will have to proxy info back into the host OS (with something like virtio-vsock or virtio-serial) whereupon QEMU and/or libvirt can pick it up and forward it on the real attestation service.

IOW, the architectural concerns raised here for SEV and SEV-ES wrt whether libvirt has ability to connect to a remote attestation server might be possible to avoid by KubeVirt taking more ownership of the exchange. When it comes to SEV-SNP though, KubeVirt's design might be forced, by the need to proxy from the guest back to the host. IOW the container in which libvirtd/qemu are running might have no choice but to provide some mechanism to connect back to an outside service.

NB, nothing about SEV-SNP is finalized, so this is just my best guess, but I feel we'll get something similar to this.

Reply:
Would it make sense to support both approaches? I dont see big conflict here. So KubeVirt may allow to do the attestation using the exported APIs as initially proposed in this PR (perhaps with a controller on top to automate the process and implement the `approve` flow). I think the use case mentioned here with bootstraping of a confidential VM with an attestation service will require smth like that. There might be other use cases as well. On the other hand, as soon as the proposed REST protocol is defined and the relevant bits are implemented in libvirt we could still support it in KubeVirt. That should give the users certain flexibility on how they do the attestation by also allowing them to bring in a 'standard' attestation service which follows that protocol.